### PR TITLE
fix: Add server checks for ticket price

### LINKED ISSUE #6280: Server checks for ticket prices are missing
**Describe the bug**
Server checks are required for ticket prices. This would also allow decimal values greater than zero to pass through [similar to eventbrite]

**Additional context**
Corresponding FE PR which requires these checks : https://github.com/fossasia/open-event-frontend/pull/3341

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Fixes #6280 

#### Short description of what this resolves:
Adds api checks for ticket prices on the server

#### Changes proposed in this pull request:

- Added check at `before_create_object` & `before_update_object`

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

File: app/api/tickets.py [hunk 1]
```diff
@@ -181,6 +184,9 @@ def before_update_object(self, ticket, data, view_kwargs):
             if not event.is_payment_enabled():
                 raise UnprocessableEntity(
                     {'event_id': ticket.event.id}, "Event having paid ticket must have a payment method")
+            if data.get('type') == 'paid' and data.get('price') <= 0:
+                raise UnprocessableEntity(
+                    {'price': data['price']}, "Price of a paid ticket must be greater than zero")
```
Comment (iamareebjamal) [inline app/api/tickets.py]: Duplicate code

File: app/api/tickets.py [hunk 2]
```diff
@@ -59,6 +66,7 @@ def before_create_object(self, data, view_kwargs):
             if not event.is_payment_enabled():
                 raise UnprocessableEntity(
                     {'event_id': data['event']}, "Event having paid ticket must have a payment method")
+        validate_ticket_price(data.get('type'), data['price'])
```
Comment (shreyanshdwivedi) [inline app/api/tickets.py]: You need to pass call this function only if `data.get('price')` is not `None` @mrsaicharan1

File: app/api/tickets.py [hunk 3]
```diff
@@ -20,6 +20,13 @@
 from app.api.helpers.exceptions import ConflictException, MethodNotAllowed, UnprocessableEntity
 from app.api.helpers.db import get_count
 
+
+def validate_ticket_price(ticket_type, price):
+    if ticket_type != 'free' and price <= 0:
```
Comment (shreyanshdwivedi) [inline app/api/tickets.py]: This check is somewhat buggy. It'll pass it ticket type is free and price < 0

Comment (mrsaicharan1) [reply app/api/tickets.py]: Yeah @shreyanshdwivedi.  Removed it.  please have a look now.

**Comment by mrsaicharan1** [comment]
@shreyanshdwivedi @iamareebjamal @uds5501 Please have a look.

File: app/api/tickets.py [hunk 4]
```diff
@@ -181,6 +189,8 @@ def before_update_object(self, ticket, data, view_kwargs):
             if not event.is_payment_enabled():
                 raise UnprocessableEntity(
                     {'event_id': ticket.event.id}, "Event having paid ticket must have a payment method")
+        if data.get('price') and data.get('type'):
```
Comment (iamareebjamal) [inline app/api/tickets.py]: Move this check inside the function

File: app/api/tickets.py [hunk 5]
```diff
@@ -22,7 +22,7 @@
 
 
 def validate_ticket_price(data):
-    if not data.get('price') and data.get('type'):
+    if not data.get('price') or not data.get('type'):
```
Comment (iamareebjamal) [inline app/api/tickets.py]: why?

**Comment by mrsaicharan1** [comment]
@iamareebjamal The DELETE test for the tickets was failing as the price & type are not required for a DELETE request. Any suggestions on how I could go about this?

**Comment by mrsaicharan1** [comment]
According to me, I think that that the condition 
```
if not data.get('price') or not data.get('type'):
        raise UnprocessableEntity({}, "Type/price of ticket is missing")
```
 should only be present under `before_create_object` as the updation of a ticket might not always involve `type` *and* `price`. What do you think?

**Comment by shreyanshdwivedi** [comment]
I agree @mrsaicharan1 
Someone can update only name or any other attribute and this check will create error

**Comment by mrsaicharan1** [comment]
@iamareebjamal @uds5501 @shreyanshdwivedi Fixed the check to be compatible with the DELETE operation as well. Ready for another review.

**Comment by mrsaicharan1** [comment]
@iamareebjamal This fix seems to be buggy with donation tickets. I will revert this for now and see what's wrong.

**Pull request merged**